{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "40d39202-2d16-42aa-b3a4-dfd8a06de354",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "name of file DNA_code.txt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCTGTGCTGACTCCCGGTGCTGCCGCTGCCATAGCTAAAGCCCGGGTCCTGGTAGGCAGGCGGGAAGCAGGGTGGGGGTCCCGGGTACTGGTAGGGGTAGCCCTGACCCAGAGGCGGGGGGCAGCCGGGTGGGGCAGCGGGGCCAGCGTGTCCTGAA-CGAAGTCCCACTGGAGCCACTGTTGAGGTTCAGGGTGGCGAGATCTGGCGGNNNAGGGTAGGTGAGGGCCGCGGAGGGGCCTCCGGCGTTCCCCTCCCCCCCGCCCTGAAACCCGAAGCCCCCACTCACTGCTGCAGAGATCCCCTGAAAACGTAGTAGCACTGCTCgagacAGGTGATCTGTTGACCTGAAACCGCAGGAAGCCGTGCTTCAGCAAGCTGCTGGCGTACTTCCGGGCCT---GCCGCTCCTTGAAGCCCTCCACGTGTGTGTACAGCCAGTCCACCACGTCCGCCCCTGGCCGGCACCAGCGGTCAGCCCGCAGCCTCGAGGCAAGCAGCCCTGCCNNTGGCACTATCCGC-CGCGGGGACGGCCACTCACCGATGACGGCATNNGCGATGGTGATCTTGAGCCACATGCGGTCGCGGATCTCCAGTCCCGAG---GGCAGCTGCATGACCCGGACGACGGCGCTCATGTCACtcaccgtcagcggcgcctcttccagCCAGCTCTGCAAAGCACAGACAGCCCCGCTTCGCCCCAGCATCTGAAAGCGGGGGACTCggcAcgCTGCACCCCCAGGGGAGCCTCTGGGCAGAGCCTGCGCCAGGGCGCAAGCTGGACGGTGCGTGACAGCAGGGAATTCCGCCACTGCAGGATGCACCCCCGTGAGGCTGGGGCGTGAGCAGGGGGGTTGGACAtttAGTCTCCCGAATTCACAGACACTTTTCATCAGGATCCTAGGCACAAACTGGGCTGAAACCCCACCCTGCAGACCAGGAAGTAATGAGAACAGGGCAGGCCCCTTCCCCTCNNCGCATGCC-CACCCGAGAGCGCAGGCTGTTAGTCGTGTTAATGGCAGGAAGCAGAATGGAGACCTGGCCCCTGCCTCTGAA-CCGTGGGTGCTCaactggctaGCCCTACGTACATCCCCTGTTCcggCCAACACACAGACATGAGCAGGATGGGCTGCACAAGGTGGGCACGGGTGCCTGTGCACACGTCTGTGCAGGGAGTTGGGGACAGGCAACACACACGTGTCACAGCCCCATGACGGggcaattgcGCCATGCTGGCTGAATGGCAGAGACGCCCCTCCAAGCCTCGGTTTCTGCTGGGGCCCTCAGGAGCTGCCACTTACGTGGAGCACCAGGCACGGAGCTGGTTAGTGAGGAGGAGCTGGTGCGCGTGACGGCGCTGGAGCAGGGACTCGTACCGTAGCGGGGCAGGGCNNNTGTCAGTGCCGCCGTGTGGtcagcggcgatCGGCG-GGTCGATGGGCCGCACCGGGTCAGCTGGGTGNAGACACGTGGCGATGACAGGCGGACAGATGGACAGGGTGGGAGGGCAGGGTGCAGGGCACAGAGGAGAGAGGCCTTCAGGCTAGGTAGGCGCCCCCTCCCCATCCCGccccGTGTGCCCCGAGGGCCACTCACCCCGTGGGACGGTGAAGTAGCTTCG-GGCGTTGGGTCCAGCACTTGGCCACAGTGAGGCTGNAAATGGCTGCAGGAACGGTGGTCCCCCCGCAAGGCCCCCATGGTCCCACCTCCCTGCCTGGCCCCTCCCGCTCCAGCGCCNCCAGCC\n"
     ]
    }
   ],
   "source": [
    "# -*- coding: utf8 -*-\n",
    "'''wczytanie pliku z formacie TXT/FASTA'''\n",
    "import sys\n",
    "from functools import partial\n",
    "import pandas as pd\n",
    "from collections import Counter\n",
    "\n",
    "#pozostanie, jeśli pojawi sie konwersja do Data frame\n",
    "if len(sys.argv) == 1:\n",
    "    print('Please provide a command line argument as a file name!')\n",
    "    sys.exit()\n",
    "else:\n",
    "    myfile = sys.argv[1]\n",
    "\n",
    "sequence = ''\n",
    "\n",
    "try:\n",
    "    with open(input('name of file'), 'r') as f:\n",
    "        for line in f:\n",
    "            if '>' not in line:\n",
    "                sequence = sequence + line.strip()\n",
    "except OSError as oserr:\n",
    "    print('OS error:', oserr)\n",
    "else:\n",
    "    print(sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6927d933-c48a-4665-9417-9dbb579ef717",
   "metadata": {},
   "outputs": [],
   "source": [
    "codons_full = {\n",
    "\"UUU\" : \"Phenylalanine\",\n",
    "\"CUU\" : \"Leucine\",\n",
    "\"AUU\" : \"Isoleucine\",\n",
    "\"GUU\" : \"Valine\",\n",
    "\"UUC\" : \"Phenylalanine\",\n",
    "\"CUC\" : \"Leucine\",\n",
    "\"AUC\" : \"Isoleucine\",\n",
    "\"GUC\" : \"Valine\",\n",
    "\"UUA\" : \"Leucine\",\n",
    "\"CUA\" : \"Leucine\",\n",
    "\"AUA\" : \"Isoleucine\",     \n",
    "\"GUA\" : \"Valine\",\n",
    "\"UUG\" : \"Leucine\",\n",
    "\"CUG\" : \"Leucine\",\n",
    "\"AUG\" : \"Methionine\",\n",
    "\"GUG\" : \"Valine\",\n",
    "\"UCU\" : \"Serine\",\n",
    "\"CCU\" : \"Proline\",\n",
    "\"ACU\" : \"Threonine\",\n",
    "\"GCU\" : \"Alanine\",\n",
    "\"UCC\" : \"Serine\",\n",
    "\"CCC\" : \"Proline\",\n",
    "\"ACC\" : \"Threonine\",\n",
    "\"GCC\" : \"Alanine\",\n",
    "\"UCA\" : \"Serine\",\n",
    "\"CCA\" : \"Proline\",\n",
    "\"ACA\" : \"Threonine\",\n",
    "\"GCA\" : \"Alanine\",\n",
    "\"UCG\" : \"Serine\",\n",
    "\"CCG\" : \"Proline\",\n",
    "\"ACG\" : \"Threonine\",\n",
    "\"GCG\" : \"Alanine\",\n",
    "\"UAU\" : \"Tyrosine\",\n",
    "\"CAU\" : \"Histidine\",\n",
    "\"AAU\" : \"Asparagine\",\n",
    "\"GAU\" : \"Aspartic acid\",\n",
    "\"UAC\" : \"Tyrosine\",\n",
    "\"CAC\" : \"Histidine\",\n",
    "\"AAC\" : \"Asparagine\",\n",
    "\"GAC\" : \"Aspartic acid\",\n",
    "\"UAA\" : \"STOP\",\n",
    "\"CAA\" : \"Glutamine\",\n",
    "\"AAA\" : \"Lysine\",\n",
    "\"GAA\" : \"Glutamic acid\",\n",
    "\"UAG\" : \"STOP\",\n",
    "\"CAG\" : \"Glutamine\",\n",
    "\"AAG\" : \"Lysine\",\n",
    "\"GAG\" : \"Glutamic acid\",\n",
    "\"UGU\" : \"Cysteine\",\n",
    "\"CGU\" : \"Arginine\",\n",
    "\"AGU\" : \"Serine\",\n",
    "\"GGU\" : \"Glycine\",\n",
    "\"UGC\" : \"Cysteine\",\n",
    "\"CGC\" : \"Arginine\",\n",
    "\"AGC\" : \"Serine\",\n",
    "\"GGC\" : \"Glycine\",\n",
    "\"UGA\" : \"STOP\",\n",
    "\"CGA\" : \"Arginine\",\n",
    "\"AGA\" : \"Arginine\",\n",
    "\"GGA\" : \"Glycine\",\n",
    "\"UGG\" : \"Tryptophan\",\n",
    "\"CGG\" : \"Arginine\",\n",
    "\"AGG\" : \"Arginine\",\n",
    "\"GGG\" : \"Glycine\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ca0c996b-1d94-4980-a423-f566c325fd36",
   "metadata": {},
   "outputs": [],
   "source": [
    "codons_short = {\n",
    "\"UUU\" : \"F\",\n",
    "\"CUU\" : \"L\",\n",
    "\"AUU\" : \"I\",\n",
    "\"GUU\" : \"V\",\n",
    "\"UUC\" : \"F\",\n",
    "\"CUC\" : \"L\",\n",
    "\"AUC\" : \"I\",\n",
    "\"GUC\" : \"V\",\n",
    "\"UUA\" : \"L\",\n",
    "\"CUA\" : \"L\",\n",
    "\"AUA\" : \"I\",     \n",
    "\"GUA\" : \"V\",\n",
    "\"UUG\" : \"L\",\n",
    "\"CUG\" : \"L\",\n",
    "\"AUG\" : \"M\",\n",
    "\"GUG\" : \"V\",\n",
    "\"UCU\" : \"S\",\n",
    "\"CCU\" : \"P\",\n",
    "\"ACU\" : \"T\",\n",
    "\"GCU\" : \"A\",\n",
    "\"UCC\" : \"S\",\n",
    "\"CCC\" : \"P\",\n",
    "\"ACC\" : \"T\",\n",
    "\"GCC\" : \"A\",\n",
    "\"UCA\" : \"S\",\n",
    "\"CCA\" : \"P\",\n",
    "\"ACA\" : \"T\",\n",
    "\"GCA\" : \"A\",\n",
    "\"UCG\" : \"S\",\n",
    "\"CCG\" : \"P\",\n",
    "\"ACG\" : \"T\",\n",
    "\"GCG\" : \"A\",\n",
    "\"UAU\" : \"Y\",\n",
    "\"CAU\" : \"H\",\n",
    "\"AAU\" : \"N\",\n",
    "\"GAU\" : \"D\",\n",
    "\"UAC\" : \"Y\",\n",
    "\"CAC\" : \"H\",\n",
    "\"AAC\" : \"N\",\n",
    "\"GAC\" : \"D\",\n",
    "\"UAA\" : \"STOP\",\n",
    "\"CAA\" : \"Q\",\n",
    "\"AAA\" : \"K\",\n",
    "\"GAA\" : \"E\",\n",
    "\"UAG\" : \"STOP\",\n",
    "\"CAG\" : \"Q\",\n",
    "\"AAG\" : \"K\",\n",
    "\"GAG\" : \"E\",\n",
    "\"UGU\" : \"C\",\n",
    "\"CGU\" : \"R\",\n",
    "\"AGU\" : \"S\",\n",
    "\"GGU\" : \"G\",\n",
    "\"UGC\" : \"C\",\n",
    "\"CGC\" : \"R\",\n",
    "\"AGC\" : \"S\",\n",
    "\"GGC\" : \"G\",\n",
    "\"UGA\" : \"STOP\",\n",
    "\"CGA\" : \"R\",\n",
    "\"AGA\" : \"R\",\n",
    "\"GGA\" : \"G\",\n",
    "\"UGG\" : \"W\",\n",
    "\"CGG\" : \"R\",\n",
    "\"AGG\" : \"R\",\n",
    "\"GGG\" : \"G\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "80c4fc70-7f5a-497f-95af-ed0e0eaa7df5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def base_DNA_cleaning(sequence):\n",
    "    \"\"\"Function that cleans the sequence from '-|N' and white spaces. Moreover, function makes sure \n",
    "    that the returned sequence is in capital letters.\n",
    "    \n",
    "    :param seq: str - sequence of DNA or RNA.\n",
    "    :return: str - cleaned DNA or RNA.\n",
    "    \"\"\"\n",
    "\n",
    "    sequence = sequence.upper()\n",
    "    if sequence in ['-', 'N', ' ']:\n",
    "        return ''\n",
    "    elif sequence == 'U':\n",
    "        return 'U'\n",
    "    elif sequence == 'A': \n",
    "        return 'A'\n",
    "\n",
    "    elif sequence == 'T':\n",
    "        return 'T'\n",
    "\n",
    "    elif sequence == 'C':\n",
    "        return 'C'\n",
    "\n",
    "    elif sequence == 'G':\n",
    "        return 'G'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "71d301ff-849d-4f02-8c13-e64a6e66c221",
   "metadata": {},
   "outputs": [],
   "source": [
    "def ATGC_count(sequence, rounding=None):\n",
    "    \"\"\"Function that counts the amount of certain nitrogenous principles and GC %, if u added rounding parameter, for default its None.\n",
    "    \n",
    "    :param sequence: str - sequence of DNA.\n",
    "    :param rounding: int - determines the approximation. Default specified as None.\n",
    "    :return: dictionary - amount of nitrogenous principles.\n",
    "    \"\"\"\n",
    "    dictionary = dict(Counter(sequence))\n",
    "    if rounding != None:\n",
    "        perc = ((dictionary.get('G') + dictionary.get('C')) * 100) / sum(dictionary.values())\n",
    "        perc = round(perc, int(rounding))\n",
    "        dictionary.update({'%GC': perc})\n",
    "    return dictionary\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "35bb0fb7-e2ac-4581-ab02-88650e61a3ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def to_rna_element(sequence, seq=None):\n",
    "    \"\"\"Function that mainly transtripts DNA strand into RNA strand. There is one more function.\n",
    "    If u add \"DNA\" as seq parameter, it will return DNA complementary to given one, not RNA.\n",
    "\n",
    "    :param sequence: str - string containing the DNA strand.\n",
    "    :param seq: str - its additional param. Can be None or DNA, others won't do anything special.\n",
    "    :return: str - transcripted RNA/DNA string.\n",
    "    \"\"\"\n",
    "    if (sequence == 'A') and (seq == 'DNA'):\n",
    "        return 'T'\n",
    "\n",
    "    elif (sequence == 'A') and (seq != 'DNA'): \n",
    "        return 'U'\n",
    "\n",
    "    elif sequence == 'T':\n",
    "        return 'A'\n",
    "\n",
    "    elif sequence == 'C':\n",
    "        return 'G'\n",
    "\n",
    "    elif sequence == 'G':\n",
    "        return 'C'\n",
    "    else:\n",
    "        raise ValueError(\"Probably your DNA strand needs some cleaning. There is something different from 'A|T|G|C' \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c8605796-bb67-4dc5-93ea-7fc26314d501",
   "metadata": {},
   "outputs": [],
   "source": [
    "def proteins(sequence, lenght=None):\n",
    "    '''This function provide Amino acids sequence in return based on given RNA sequence. \n",
    "    The translation is being stopped, when it hits one of STOP codons. The sequence does not need to be started\n",
    "    with \"AUG\".\n",
    "    \n",
    "    :param strand: str - RNA sequence.\n",
    "    :param lenght: str - 'short' or 'long', it determines whatever the output amino-acids are in long, full name like \n",
    "    \"Methionine\" or in short one like \"M\".\n",
    "    :return: str - the sequence of amino-acids.\n",
    "    '''\n",
    "    \n",
    "    codons = None\n",
    "    separator = ''\n",
    "    proteins = []\n",
    "    if 'T' in sequence:\n",
    "        raise ValueError('You provided DNA sequence instead of RNA sequence')\n",
    "        \n",
    "    if lenght == 'long':\n",
    "        codons = codons_full\n",
    "        separator = ', '\n",
    "    elif (lenght == 'short') or (lenght == None):\n",
    "        codons = codons_short  \n",
    "    else:\n",
    "        raise ValueError('wrong input in place of lenght. It has to be \"long\" for long name of amino-acids or \"short\" for the short ones.')\n",
    "        \n",
    "    for i in range(0, len(sequence), 3):\n",
    "        protein = codons[sequence[i:i+3]]\n",
    "        if protein == 'STOP':\n",
    "            break\n",
    "        proteins.append(protein)\n",
    "        result = separator.join(proteins)\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f85eba2a-a7ab-4c6a-9ab1-d3ca2a004568",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_pattern(sequence, pattern):\n",
    "    \"\"\"Function that check for the specific pattern in the given sequence.\n",
    "\n",
    "    :param sequence: str - sequence of DNA or RNA or other sequence.\n",
    "    :param pattern: str - pattern of the shorter sequence like 'GAATTC'.\n",
    "    :return: str - if there is occur of the pattern, it will return the beggining \n",
    "    index and final index of the pattern's nucleotides\n",
    "    \"\"\"\n",
    "    import re\n",
    "    import sys\n",
    "    try:\n",
    "        re.compile(pattern)\n",
    "    except:\n",
    "        print('Invalid pattern. Please make sure to input correct one')\n",
    "        sys.exit()\n",
    "        \n",
    "    match = re.search(pattern, sequence)\n",
    "    \n",
    "    if match:\n",
    "        print('Found the occur to the pattern   :', match.group())\n",
    "        print('Starting at index :', match.start())\n",
    "        print('Ending at index   :', match.end())\n",
    "    else:\n",
    "        print('Did not find the pattern.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f3b79e1f-f5ed-4bae-90df-0adf23a27337",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sequencing(sequence, pattern):\n",
    "    \"\"\"Function that check for the specific repetition in the given sequence.\n",
    "\n",
    "    :param sequence: str - sequence of DNA or RNA or other sequence.\n",
    "    :param pattern: str - pattern of the short sequence like 'GAATTC'.\n",
    "    :return: str - if there is at least one occur of the pattern, it will return the part of sequence from the start to the hit of pattern.\n",
    "    \"\"\"\n",
    "    \n",
    "    amount = re.findall(pattern, sequence)\n",
    "    occur = f'There is/are {len(amount)} occurencies of {pattern} found in. First on {sequence.index(pattern)}'\n",
    "    if len(amount) >= 1:\n",
    "        result = re.split(pattern, sequence)\n",
    "        for number, match in enumerate(result[:-1]):\n",
    "            print(f' Wzór sekwencji {number + 1}: {match + pattern}')\n",
    "            print()\n",
    "    return occur"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f154a570-04d7-4b9c-aaaa-519d49c8c037",
   "metadata": {},
   "outputs": [],
   "source": [
    "def helper(action = None):\n",
    "    '''Function that provides help for user.\n",
    "    \n",
    "    :param action: str - the function number, for which the information will be provided.\n",
    "    :return: str - information about specific function. \n",
    "    '''\n",
    "    if action == '1':\n",
    "        print(base_DNA_cleaning.__doc__)\n",
    "    elif action == '2':\n",
    "        print(ATGC_count.__doc__)\n",
    "    elif action == '3':\n",
    "        print(to_rna_element.__doc__)\n",
    "    elif action == '4':\n",
    "        print(proteins.__doc__)\n",
    "    elif action == '5':\n",
    "        print(find_pattern.__doc__)\n",
    "    elif action == '6':\n",
    "        print(sequencing.__doc__)\n",
    "    else:\n",
    "        print('''This is a simple program that provides an opportunity to perform several \n",
    "        operations on a sequence (mainly DNA sequence) such as:\n",
    "        1. Base DNA cleaning\n",
    "        2. A, T, G, C counting (with optional %GC)\n",
    "        3. Obtaining RNA after transcription process of DNA sequence which was provided\n",
    "        or obtaining a strand complementary to the inserted one.\n",
    "        4. Get Amino-acids sequence based on DNA \n",
    "        5. Check for the specific pattern in sequence.\n",
    "        6. Split sequences based on pattern occurrence\n",
    "\n",
    "        To get some more information about specific step, write for example(for first one): \"info 1\".\n",
    "        To get started write simply the specific number of action you want to do, for example: \"1\".\n",
    "        If you have finished your work here, type \"exit\".\n",
    "        To show the description later type \"help\".\n",
    "        To show the current sequence type \"sequence\".\n",
    "        ''')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "022cb28b-a8a8-487b-b8d7-ea8c3bfad25f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "choose an action to do: \n",
      " 1\n",
      "Do you want to save it as new file? type yes or no: \n",
      " 1\n",
      "choose an action to do: \n",
      " 1\n",
      "Do you want to save it as new file? type yes or no: \n",
      " no\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCTGTGCTGACTCCCGGTGCTGCCGCTGCCATAGCTAAAGCCCGGGTCCTGGTAGGCAGGCGGGAAGCAGGGTGGGGGTCCCGGGTACTGGTAGGGGTAGCCCTGACCCAGAGGCGGGGGGCAGCCGGGTGGGGCAGCGGGGCCAGCGTGTCCTGAACGAAGTCCCACTGGAGCCACTGTTGAGGTTCAGGGTGGCGAGATCTGGCGGAGGGTAGGTGAGGGCCGCGGAGGGGCCTCCGGCGTTCCCCTCCCCCCCGCCCTGAAACCCGAAGCCCCCACTCACTGCTGCAGAGATCCCCTGAAAACGTAGTAGCACTGCTCGAGACAGGTGATCTGTTGACCTGAAACCGCAGGAAGCCGTGCTTCAGCAAGCTGCTGGCGTACTTCCGGGCCTGCCGCTCCTTGAAGCCCTCCACGTGTGTGTACAGCCAGTCCACCACGTCCGCCCCTGGCCGGCACCAGCGGTCAGCCCGCAGCCTCGAGGCAAGCAGCCCTGCCTGGCACTATCCGCCGCGGGGACGGCCACTCACCGATGACGGCATGCGATGGTGATCTTGAGCCACATGCGGTCGCGGATCTCCAGTCCCGAGGGCAGCTGCATGACCCGGACGACGGCGCTCATGTCACTCACCGTCAGCGGCGCCTCTTCCAGCCAGCTCTGCAAAGCACAGACAGCCCCGCTTCGCCCCAGCATCTGAAAGCGGGGGACTCGGCACGCTGCACCCCCAGGGGAGCCTCTGGGCAGAGCCTGCGCCAGGGCGCAAGCTGGACGGTGCGTGACAGCAGGGAATTCCGCCACTGCAGGATGCACCCCCGTGAGGCTGGGGCGTGAGCAGGGGGGTTGGACATTTAGTCTCCCGAATTCACAGACACTTTTCATCAGGATCCTAGGCACAAACTGGGCTGAAACCCCACCCTGCAGACCAGGAAGTAATGAGAACAGGGCAGGCCCCTTCCCCTCCGCATGCCCACCCGAGAGCGCAGGCTGTTAGTCGTGTTAATGGCAGGAAGCAGAATGGAGACCTGGCCCCTGCCTCTGAACCGTGGGTGCTCAACTGGCTAGCCCTACGTACATCCCCTGTTCCGGCCAACACACAGACATGAGCAGGATGGGCTGCACAAGGTGGGCACGGGTGCCTGTGCACACGTCTGTGCAGGGAGTTGGGGACAGGCAACACACACGTGTCACAGCCCCATGACGGGGCAATTGCGCCATGCTGGCTGAATGGCAGAGACGCCCCTCCAAGCCTCGGTTTCTGCTGGGGCCCTCAGGAGCTGCCACTTACGTGGAGCACCAGGCACGGAGCTGGTTAGTGAGGAGGAGCTGGTGCGCGTGACGGCGCTGGAGCAGGGACTCGTACCGTAGCGGGGCAGGGCTGTCAGTGCCGCCGTGTGGTCAGCGGCGATCGGCGGGTCGATGGGCCGCACCGGGTCAGCTGGGTGAGACACGTGGCGATGACAGGCGGACAGATGGACAGGGTGGGAGGGCAGGGTGCAGGGCACAGAGGAGAGAGGCCTTCAGGCTAGGTAGGCGCCCCCTCCCCATCCCGCCCCGTGTGCCCCGAGGGCCACTCACCCCGTGGGACGGTGAAGTAGCTTCGGGCGTTGGGTCCAGCACTTGGCCACAGTGAGGCTGAAATGGCTGCAGGAACGGTGGTCCCCCCGCAAGGCCCCCATGGTCCCACCTCCCTGCCTGGCCCCTCCCGCTCCAGCGCCCCAGCC\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "choose an action to do: \n",
      " exit\n"
     ]
    }
   ],
   "source": [
    "action = input('choose an action to do: \\n')\n",
    "while action != 'exit':\n",
    "    if action.startswith('info'):\n",
    "        helper(action[-1])\n",
    "    elif action == 'help':\n",
    "        helper()\n",
    "    elif action == 'sequence':\n",
    "        print(sequence)\n",
    "    \n",
    "    \n",
    "    elif action == '1':\n",
    "        sequence = ''.join(map(base_DNA_cleaning, sequence))\n",
    "        save = input('Do you want to save it as new file? type yes or no: \\n')\n",
    "        if save.lower() == 'yes': \n",
    "            name = input('name of file: \\n')\n",
    "            with open(name, 'w') as f:\n",
    "                f.write(sequence)\n",
    "        elif save.lower() == 'no':\n",
    "            print(sequence)\n",
    "    elif action == '2':\n",
    "        input_text = input('if you want to get %GC, type numerical approximation (like \"2\"), if not, press enter: \\n')\n",
    "        if input_text.isdecimal():\n",
    "            rounding = int(input_text)\n",
    "            print(ATGC_count(sequence, rounding))\n",
    "        elif input_text == '':\n",
    "            print(ATGC_count(sequence))\n",
    "        else:\n",
    "            print(\"The provided value is neither a number nor enter.\\n Please select action again\")\n",
    "    elif action == '3':\n",
    "        seq = input('if you want to get complementary DNA strand, type \"DNA\", press enter to get RNA: \\n')\n",
    "        to_rna_element_2 = partial(to_rna_element, seq=seq)\n",
    "        print(''.join(map(to_rna_element_2, sequence)))\n",
    "        question = input('Do you want to keep new RNA sequence as sequence for further actions? \"yes\" or \"no\"')\n",
    "        if question.lower() == \"yes\":\n",
    "            sequence = ''.join(map(to_rna_element_2, sequence))\n",
    "    elif action == '4':\n",
    "        lenght = input('type \"short\" for single letter AA, \"long\" for full name: \\n')\n",
    "        print(proteins(sequence, lenght))\n",
    "    elif action == '5':\n",
    "        pattern = input('Type the pattern you are looking for: \\n')\n",
    "        print(find_pattern(sequence, pattern))\n",
    "    elif action == '6':\n",
    "        pattern = input('Type the pattern you are looking for: \\n')\n",
    "        print(sequencing(sequence, pattern))\n",
    "    else:\n",
    "        print('I do not understand You.')\n",
    "    action = input('choose an action to do: \\n')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
